File: SignalServiceKit/src/Network/API/Requests/TSRequest.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "TSRequest.h"
#import "TSAccountManager.h"
#import <SignalCoreKit/NSData+OWS.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@implementation TSRequest

@synthesize authUsername = _authUsername;
@synthesize authPassword = _authPassword;

- (id)initWithURL:(NSURL *)URL {
    OWSAssertDebug(URL);
    self = [super initWithURL:URL
                  cachePolicy:NSURLRequestReloadIgnoringLocalAndRemoteCacheData
              timeoutInterval:OWSRequestFactory.textSecureHTTPTimeOut];
    if (!self) {
        return nil;
    }

    _parameters = @{};
    self.shouldHaveAuthorizationHeaders = YES;

    return self;
}

- (instancetype)init
{
    OWSFail(@"You must use the initWithURL: method");
    return nil;
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wobjc-designated-initializers"

- (instancetype)initWithURL:(NSURL *)URL
                cachePolicy:(NSURLRequestCachePolicy)cachePolicy
            timeoutInterval:(NSTimeInterval)timeoutInterval
{
    OWSFail(@"You must use the initWithURL: method");
    return nil;
}

- (instancetype)initWithURL:(NSURL *)URL
                     method:(NSString *)method
                 parameters:(nullable NSDictionary<NSString *, id> *)parameters
{
    OWSAssertDebug(URL);
    OWSAssertDebug(method.length > 0);

    self = [super initWithURL:URL
                  cachePolicy:NSURLRequestReloadIgnoringLocalAndRemoteCacheData
              timeoutInterval:OWSRequestFactory.textSecureHTTPTimeOut];
    if (!self) {
        return nil;
    }

    _parameters = parameters ?: @{};
    [self setHTTPMethod:method];
    self.shouldHaveAuthorizationHeaders = YES;

    return self;
}

+ (instancetype)requestWithUrl:(NSURL *)url
                        method:(NSString *)method
                    parameters:(nullable NSDictionary<NSString *, id> *)parameters
{
    return [[TSRequest alloc] initWithURL:url method:method parameters:parameters];
}

#pragma mark - Authorization

- (void)setAuthUsername:(nullable NSString *)authUsername
{
    OWSAssertDebug(self.shouldHaveAuthorizationHeaders);

    @synchronized(self) {
        _authUsername = authUsername;
    }
}

- (void)setAuthPassword:(nullable NSString *)authPassword
{
    OWSAssertDebug(self.shouldHaveAuthorizationHeaders);

    @synchronized(self) {
        _authPassword = authPassword;
    }
}

- (nullable NSString *)authUsername
{
    OWSAssertDebug(self.shouldHaveAuthorizationHeaders);

    @synchronized(self) {
        NSString *_Nullable result = (_authUsername ?: self.tsAccountManager.storedServerUsername);
        if (result.length < 1) {
            OWSLogVerbose(@"%@", self.debugDescription);
        }
        OWSAssertDebug(result.length > 0);
        return result;
    }
}

- (nullable NSString *)authPassword
{
    OWSAssertDebug(self.shouldHaveAuthorizationHeaders);

    @synchronized(self) {
        NSString *_Nullable result = (_authPassword ?: self.tsAccountManager.storedServerAuthToken);
        if (result.length < 1) {
            OWSLogVerbose(@"%@", self.debugDescription);
        }
        OWSAssertDebug(result.length > 0);
        return result;
    }
}

- (NSString *)description {
    if (self.shouldRedactUrlInLogs) {
        return [NSString stringWithFormat:@"{ %@: [REDACTED] }", self.HTTPMethod];
    } else {
        return [NSString stringWithFormat:@"{ %@: %@ }", self.HTTPMethod, self.URL];
    }
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Network/API/Requests/TSRequest.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

// TODO: Rework to _not_ extend NSMutableURLRequest.
@objcMembers
public class TSRequest: NSMutableURLRequest {
    public var isUDRequest: Bool = false
    public var shouldHaveAuthorizationHeaders: Bool = true

    /// If true, an HTTP 401 will trigger a follow up request to see if the account is deregistered.
    /// If it is, the account will be marked as de-registered.
    ///
    /// - Warning: This only applies to REST requests. We handle HTTP 403 errors
    /// (*not* HTTP 401) for web sockets during the initial handshake, not
    /// during the processing for individual requests.
    public var shouldCheckDeregisteredOn401: Bool = false

    public let parameters: [String: Any]

    public init(url: URL) {
        parameters = [:]
        super.init(
            url: url,
            cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
            timeoutInterval: OWSRequestFactory.textSecureHTTPTimeOut
        )
    }

    public init(url: URL, method: String, parameters: [String: Any]?) {
        owsAssertDebug(method.isEmpty.negated)

        self.parameters = parameters ?? [:]
        super.init(
            url: url,
            cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
            timeoutInterval: OWSRequestFactory.textSecureHTTPTimeOut
        )
        self.httpMethod = method
    }

    @objc(requestWithUrl:method:parameters:)
    public static func request(url: URL, method: String, paramters: [String: Any]?) -> TSRequest {
        return TSRequest(url: url, method: method, parameters: paramters)
    }

    @available(*, unavailable)
    public override init(url: URL, cachePolicy: NSURLRequest.CachePolicy, timeoutInterval: TimeInterval) {
        fatalError()
    }

    public required init?(coder: NSCoder) {
        fatalError("init(coder:) unavailable")
    }

    // MARK: - Authorization

    private let authLock = UnfairLock()

    private var _authUsername: String?
    public var authUsername: String? {
        get {
            owsAssertDebug(shouldHaveAuthorizationHeaders)
            return authLock.withLock {
                let result = _authUsername ?? self.tsAccountManager.storedServerUsername
                if result.isEmptyOrNil {
                    Logger.verbose(self.debugDescription)
                }
                owsAssertDebug(result.isEmptyOrNil.negated)
                return result
            }
        }
        set {
            owsAssertDebug(shouldHaveAuthorizationHeaders)
            authLock.withLock {
                _authUsername = newValue
            }
        }
    }

    private var _authPassword: String?
    public var authPassword: String? {
        get {
            owsAssertDebug(shouldHaveAuthorizationHeaders)
            return authLock.withLock {
                let result = _authPassword ?? self.tsAccountManager.storedServerAuthToken
                if result.isEmptyOrNil {
                    Logger.verbose(self.debugDescription)
                }
                owsAssertDebug(result.isEmptyOrNil.negated)
                return result
            }
        }
        set {
            owsAssertDebug(shouldHaveAuthorizationHeaders)
            authLock.withLock {
                _authPassword = newValue
            }
        }
    }

    public enum RedactionStrategy {
        case none
        /// Error responses must be separately handled
        case redactURLForSuccessResponses(replacementString: String = "[REDACTED]")
    }

    private var redactionStrategy = RedactionStrategy.none

    public func applyRedactionStrategy(_ strategy: RedactionStrategy) {
        self.redactionStrategy = strategy
    }

    public func objc_applySuccessResponsesURLRedactionStrategy() {
        self.applyRedactionStrategy(.redactURLForSuccessResponses())
    }

    public override var description: String {
        switch redactionStrategy {
        case .none:
            return "{ \(self.httpMethod): \(String(describing: self.url)) }"
        case .redactURLForSuccessResponses(let replacementString):
            return "{ \(self.httpMethod): \(replacementString) }"
        }
    }
}
